### ISSUE TITLE
Let nats client send auth_required without waiting for the server to ask for it

### ISSUE THREAD

**Issue by lubien** [nats-io/nats.ex#132]
I've been using a [fork of your library](https://github.com/nats-io/nats.ex/compare/main...lubien:nats.ex:main) on [Fly.io for logs](https://community.fly.io/t/fly-logs-over-nats/1540) with a few changes.

My proposal negotiate auth if the client wrote those on its config without needing for the server to ask for it. I've implemented here an example: https://github.com/nats-io/nats.ex/commit/9ddb11acbca7727e97c6624425a2cb22243cfafe

This feels like a breaking change for some so I'd love input on this!

**Comment by mmmries** [comment]
Hmmmm I've never thought about forcing a connection to use auth. Is that supported in the protocol? I thought we always did an initial handshake first and we could only send the credentials as our response to that handshake?

**Comment by mmmries** [comment]
Looking at the [client protocol](https://docs.nats.io/reference/reference-protocols/nats-protocol#connect) it indicates that `auth_token`, `user`, or `pass` should only be sent if `auth_required` is set. Do other nats clients allow the client for force authentication?

If we did want to support this, I think I would avoid modifying the `server_settings` and instead I would handle this logic in the `negotiate_settings` function. So that function will always see the server settings and user_settings in a consistent way.

**Comment by lubien** [comment]
> Do other nats clients allow the client for force authentication?

Yes, Go client does: https://github.com/nats-io/nats.go/blob/main/nats.go#L2248

Which we happen to use on our CLI: https://github.com/superfly/flyctl/blob/master/logs/nats.go#L78

Since our NATs server does not ask for `auth_required`.



> If we did want to support this, I think I would avoid modifying the `server_settings` and instead I would handle this logic in the `negotiate_settings` function. So that function will always see the server settings and user_settings in a consistent way.

Got it. I can take a stab at this if this is wanted.

**Comment by mmmries** [comment]
Hey @lubien thank you for opening the PR. Can you clarify a few things for me?

> Yes, Go client does: https://github.com/nats-io/nats.go/blob/main/nats.go#L2248

I'm not very familiar with Go, but it looks like that code will sent the auth credentials as long as they have been provided by the user in the `nc.Opts`? Is that right? So I wonder if we should change the Elixir client to follow that same convention and skip checking the `server_settings.auth_required`? We would just send the credentials if the use has specified them?

> Since our NATs server does not ask for auth_required.

So your NATS server is configured in a way that it does not send `"auth_required": true` in the INFO message, but it *does* actually require credentials in order to connect successfully? I'm happy to follow the conventions in the go client, I just want to make sure I'm understanding how your NATS server is setup.

**Comment by lubien** [comment]
> will sent the auth credentials as long as they have been provided by the user in the `nc.Opts`?

Yes, that's my understanding too. Disclaimer: I'm very new to Go 😆 

> So your NATS server is configured in a way that it does not send `"auth_required": true` in the INFO message, but it _does_ actually require credentials in order to connect successfully?

Yes, that's how we built that. We just expect it to be sent.

Thanks for your time, really :)

**Comment by mmmries** [comment]
Hi @pguillory thanks for nudging this. I hadn't heard any more use-cases around this since @lubien first identified it and proposed a fix. I did some more digging on this and it seems like it is a known edge case having to do with multi-account configurations in the server (see https://github.com/nats-io/nats-server/discussions/4535)

In essence it means that the server doesn't know that auth is required at the beginning of the handshake because it doesn't know which user/account you're trying to connect to. It looks to me like the go client is designed to basically ignore the `auth_required` flag and it just always sends credentials if the user has supplied the. The python client seems to do the same thing this project does and only sends them if the server sets the `auth_required: true` attribute in its' server info.

I think it might be best to follow the same behavior as the go client (definitely the reference for all clients IMO), but I'm a bit nervous about making a breaking change to the API of this library. Will it mess up any other existing use-cases?

Maybe the PR that lubien opened above is a good middle-way where users can configure proactively sending the credentials for these edge cases?
